broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/ResetMessageCorrelationProcessor.java

```diff
@@ -62,20 +70,30 @@ public void processRecord(
     }
     messageState.removeMessageCorrelation(messageKey, workflowInstanceKey);
 
-    final MessageSubscription subscription =
-        subscriptionState.get(
-            subscriptionRecord.getElementInstanceKey(), subscriptionRecord.getMessageName());
+    subscriptionState.visitSubscriptions(
+        subscriptionRecord.getMessageName(),
+        subscriptionRecord.getCorrelationKey(),
+        subscription -> {
+          if (!subscription.isCorrelating()) {
+            correlateMessage(streamWriter, subscription, sideEffect);
+            return false;
+          }
+          return true;
+        });
+  }
 
-    if (subscription != null) {
-      correlateMessageRecord.reset();
-      correlateMessageRecord
-          .setMessageKey(subscription.getMessageKey())
-          .setWorkflowInstanceKey(subscription.getWorkflowInstanceKey())
-          .setElementInstanceKey(subscription.getElementInstanceKey())
-          .setCorrelationKey(subscription.getCorrelationKey())
-          .setMessageName(subscription.getMessageName());
+  private void correlateMessage(
+      final TypedStreamWriter streamWriter,
+      final MessageSubscription subscription,
+      final Consumer<SideEffectProducer> sideEffect) {
+    correlateMessageRecord.reset();
+    correlateMessageRecord
+        .setMessageKey(subscription.getMessageKey())
+        .setWorkflowInstanceKey(subscription.getWorkflowInstanceKey())
+        .setElementInstanceKey(subscription.getElementInstanceKey())
+        .setCorrelationKey(subscription.getCorrelationKey())
+        .setMessageName(subscription.getMessageName());
 
-      streamWriter.appendNewCommand(MessageSubscriptionIntent.CORRELATE, correlateMessageRecord);
-    }
+    messageCorrelator.correlateNextMessage(subscription, correlateMessageRecord, sideEffect);
```

Comment:
This looks a bit odd because we already have the message to correlate with ;) I would like to have it more explicitly to be sure that we correlate the rejected message.